### PR TITLE
[compiler] Code generation infrastructure for on-demand method splitting

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by johnc1231** [comment]
Can we fix this up and try to get it in? I'm excited to do some ndarray experiments

**Comment by tpoterba** [comment]
Yes, definitely! I'll assign you as a reviewer, though I'd like to also do a quick design review with Cotton to make sure I'm not doing anything crazy.

I don't think there are any major redesigns I have in mind, barring feedback from review. I think I'll make it possible to control the default split size using a flag / env var, but other than that it works and seems to improve performance.

File: hail/src/main/scala/is/hail/expr/ir/analyses/CompilationUnitSize.scala [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package is.hail.expr.ir.analyses
+
+import is.hail.expr.ir._
+
+object CompilationUnitSize {
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/expr/ir/analyses/CompilationUnitSize.scala:5]: This seems like it's not used anywhere?

**Review by johnc1231 — CHANGES_REQUESTED** [review]
I'm not sure what CompilationUnitSize is for, but I otherwise think this looks good.

File: hail/src/main/scala/is/hail/expr/ir/analyses/CompilationUnitSize.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/analyses/CompilationUnitSize.scala:5]: oh, good catch! This was used in an earlier version.

Comment (johnc1231) [reply hail/src/main/scala/is/hail/expr/ir/analyses/CompilationUnitSize.scala:5]: Ah ok. Feel free to dismiss my review after deleting it.

**Review by johnc1231 — APPROVED** [review]

**Pull request merged**